<a href="https://colab.research.google.com/github/mongabra/AI-week_3/blob/main/Irisdataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Load the dataset

### Subtask:
Load the Iris dataset from Scikit-learn.


**Reasoning**:
Import the necessary function and load the Iris dataset.



In [ ]:
from sklearn.datasets import load_iris

iris_dataset = load_iris()

## Preprocess the data

### Subtask:
Handle potential missing values (although the Iris dataset is clean) and encode the target labels.


**Reasoning**:
Create a DataFrame and Series from the loaded Iris dataset, check for missing values, and print the target names.



In [ ]:
import pandas as pd

df = pd.DataFrame(iris_dataset.data, columns=iris_dataset.feature_names)
target = pd.Series(iris_dataset.target)

print("Missing values in DataFrame:")
print(df.isnull().sum())

print("\nTarget names:")
print(iris_dataset.target_names)

Missing values in DataFrame:
sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     0
dtype: int64

Target names:
['setosa' 'versicolor' 'virginica']


## Split the data

### Subtask:
Split the dataset into training and testing sets.


**Reasoning**:
Split the data into training and testing sets using `train_test_split` and store the results in the specified variables.



In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df, target, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (120, 4)
Shape of X_test: (30, 4)
Shape of y_train: (120,)
Shape of y_test: (30,)


## Train the model

### Subtask:
Train a decision tree classifier on the training data.


**Reasoning**:
Import the necessary class, instantiate the model, and fit it to the training data.



In [ ]:
from sklearn.tree import DecisionTreeClassifier

dt_classifier = DecisionTreeClassifier()
dt_classifier.fit(X_train, y_train)

DecisionTreeClassifier()

## Evaluate the model

### Subtask:
Predict on the test data and calculate accuracy, precision, and recall.


**Reasoning**:
Import necessary metrics, make predictions, and calculate evaluation scores.



In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score

y_pred = dt_classifier.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='weighted')
recall = recall_score(y_test, y_pred, average='weighted')

print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")

Accuracy: 1.0000
Precision: 1.0000
Recall: 1.0000


## Summary:

### Data Analysis Key Findings

*   The Iris dataset was successfully loaded and confirmed to have no missing values.
*   The target labels were found to be already numerically encoded (0, 1, 2) corresponding to 'setosa', 'versicolor', and 'virginica'.
*   The dataset was split into training (80%) and testing (20%) sets, resulting in 120 training samples and 30 testing samples.
*   A Decision Tree Classifier was trained on the training data.
*   The trained Decision Tree Classifier achieved perfect scores on the test set: Accuracy of 1.0000, Precision of 1.0000, and Recall of 1.0000.

### Insights or Next Steps

*   The decision tree model perfectly classified all instances in the test set, indicating excellent performance on this split of the Iris dataset.
*   While the current results are perfect, it would be beneficial to evaluate the model using cross-validation to get a more robust estimate of its performance across different data splits.
